Respect checkpointed boundaries when using knapsack formulation in the partitioner

fmassa opened the pull request:
When multiple checkpoint regions are back-to-back with no operations in-between, we enforce the operation at the boundary to be force-saved, see https://github.com/pytorch/pytorch/blob/7ea0da2d57107b087b38ec0b7fa47ff1ab84ee75/torch/_functorch/partitioners.py#L772-L807

When using the `memory_budget` formulation on a graph which already has AC inside, we should respect the boundaries of the AC decision (which is set to `MUST_SAVE`), and thus ban those nodes from possible recomputation.

Adding tests would be nice, but not sure what's the best way to test this right now.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @yf225 @xuanzhang816

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -1667,6 +1667,8 @@ def get_recomputable_banned_nodes(banned_nodes: List[fx.Node]) -> List[fx.Node]:
                 i.dist_from_bw < int(1e9)  # type: ignore[attr-defined]
                 and get_node_storage(i) not in input_storages
+                and i.meta.get("recompute", CheckpointPolicy.PREFER_SAVE)
+                                != CheckpointPolicy.MUST_SAVE
```
ccing @soulitzer in case you have opinions. I think this change seems reasonable. Maybe to restate: the partitioner is generally free to make its own decisions about what to save vs recompute, but it is expected to always respect save/recompute decisions from user `checkpoint` APIs. So in that sense, this PR seems like a strict improvement

soulitzer replied:
Agreed, we definitely shouldn't override user decision to save here (esp when none of these APIs are actually peak memory aware)

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -1667,6 +1667,8 @@ def get_recomputable_banned_nodes(banned_nodes: List[fx.Node]) -> List[fx.Node]:
                 # Only allow recomputing nodes that are actually required for BW
                 i.dist_from_bw < int(1e9)  # type: ignore[attr-defined]
                 and get_node_storage(i) not in input_storages
+                and i.meta.get("recompute", CheckpointPolicy.PREFER_SAVE)
```
I think a test would be very nice to have though, since this logic is pretty complicated to reason about and I imagine this won't be the last time we change it.

Just trying to brainstorm a test that could hit this case - what do you think of something like this:

```
def f(x):
    x = x.sin()
    return torch.matmul(x, x).sin()

# give the memory budget logic a value that should cause it to run,
# but not recompute the matmuls
activation_memory_budget=0.99
@torch.compile
def g(x):
    x = checkpoint(f, x)
    x = checkpoint(f, x)
    return x
```

where in my head, we are trying to get autoAC to run in the above example, but force it to realize that it needs to save the `aten.sin` intermediate that straddles the two checkpointed regions (even though it would otherwise clearly be a win to recompute it)

soulitzer replied:
Hmm sin is already not banned from recompute in the partitioner's compute-optimized settings, so the memory budget doesn't actually do the unbanning on sin (which I understand is what overrides the user checkpoint)

soulitzer replied:
We could also just chain a bunch of matmuls, which are all compute intensive. Memory-budget would choose a matmul based on flops/memory tradeoff to unban. We could use SAC to save a matmul output that is quite small which we know memory-budget would prioritize to unban.

bdhirsh replied:
that's a good point. Maybe something like this?
```
import torch
from torch.utils.checkpoint import checkpoint


# x: [1024, 64]
def f(x):
    tmp1 = torch.matmul(x, x.T) # [1024, 1024]
    tmp1 = torch.matmul(tmp1, tmp1) # [1024, 1024]
    tmp1 = torch.matmul(tmp1, tmp1) # [1024, 1024]
    out = torch.matmul(tmp1, x)  # [1024, 64]
    return out

# give the memory budget logic a value that should cause it to run,
# but not recompute the matmuls
from torch._functorch import config
config.activation_memory_budget=0.01
@torch.compile
def g(x):
    x = checkpoint(f, x)
    x = checkpoint(f, x)
    return x

x = torch.randn(64, 1024, requires_grad=True, device='cuda')
out = g(x)
```

I didn't run it with this PR, but right now I am seeing this P1778788661 as my forward graph, which looks wrong (only `primals_1` is getting saved for backward, not the output of the first checkpoint)

soulitzer replied:
Partitioner should already respect user - prefer/must recompute decision, so maybe everything being recomputed is expected?

fmassa commented on torch/_functorch/partitioners.py:
```diff
@@ -1780,8 +1780,14 @@ def get_recomputable_banned_nodes(
         ]
 
     recomputable_banned_nodes = get_recomputable_banned_nodes(banned_nodes)
-    # sort first by name, to ensure determinism when multiple nodes have same size
-    recomputable_banned_nodes = sorted(recomputable_banned_nodes, key=lambda x: x.name)
```
This is not needed anymore as `banned_nodes` is now an `OrderedSet` https://github.com/pytorch/pytorch/pull/146102

fmassa committed:
Respect checkpointed boundaries when using knapsack formulation in the partitioner (#141684)

Summary:
When multiple checkpoint regions are back-to-back with no operations in-between, we enforce the operation at the boundary to be force-saved, see https://github.com/pytorch/pytorch/blob/7ea0da2d57107b087b38ec0b7fa47ff1ab84ee75/torch/_functorch/partitioners.py#L772-L807

When using the `memory_budget` formulation on a graph which already has AC inside, we should respect the boundaries of the AC decision (which is set to `MUST_SAVE`), and thus ban those nodes from possible recomputation.

Adding tests would be nice, but not sure what's the best way to test this right now.

cc yf225  xuanzhang816


Differential Revision: D73718424

Pulled By: fmassa

bdhirsh commented on test/functorch/test_ac.py:
```diff
@@ -374,6 +375,33 @@ def call():
+        def call():
+            return g(x).sum()
+
+        eager_mem, eager_flops = get_mem_and_flops(call)
```
oh really nice idea using flops as our test for whether we are respecting recompute decisions!

bdhirsh reviewed (approved):
thanks for the test!

fmassa commented:
@pytorchbot merge
